### ISSUE TITLE
"No binding found" error by generic interface

### ISSUE THREAD

**Issue by k-kagurazaka** [SalomonBrys/Kodein#2]
Hi,

this library looks pretty good, so it's simple, powerful and quite Kotlin like.
However I encountered a problem when I try to inject a instance of generic interface.

Example code:

```kotlin
public interface Foo
public interface Bar<T>
public class FooBarImpl : Bar<Foo>

override val kodein: Kodein by lazyKodein {
    bind<Bar<Foo>>() with instance(FooBarImpl())
}

val foobar: Bar<Foo> by injectInstance() // "No binding found" exception!!
```

The instance of Kodein properly has the instance of FooBarImpl in its map, but binding failed.
I found an uncool solution using a wrapper interface:

```kotlin
public interface Foo
public interface Bar<T>
public interface FooBar : Bar<Foo>
public class FooBarImpl : FooBar

override val kodein: Kodein by lazyKodein {
    bind<FooBar>() with instance(FooBarImpl())
}

val foobar: FooBar by injectInstance()
```

It works well, but I am not a fan of the redundant interface.
Do you have any ideas for solving this issue?

**Comment by k-kagurazaka** [comment]
I found my misunderstanding.
Kodein with the generic interface works fine.
Sorry.

**Comment by SalomonBrys** [comment]
Hi, can you tell me what was your misunderstanding. Maybe I could update the doc ;)

**Comment by k-kagurazaka** [comment]
Firstly, the "No binding found" problem occurred in my Android project.
However, I tried to add a generic interface test and to run the test, as a result, all tests passed.
I thought this is a specific issue of my project, so I commented "misunderstanding".

But now, I tried to add the above "FooBar" code to the "AndroidDemo" of this repository, the same problem occurred in my KitKat device.

The code is here:
https://github.com/k-kagurazaka/Kodein/tree/generic-interface/AndroidDemo

In my lollipop device (Nexus 6), this app works correctly.
It is necessary to survey environments which the issue will occur.

**Comment by SalomonBrys** [comment]
I'm currently in vacation and I don't have access to a computer. I will look at this as soon as I can... meaning next Monday.

**Comment by SalomonBrys** [comment]
I can reproduce this, I am investigating ;)

**Comment by SalomonBrys** [comment]
OK, I found the error.
To support Specific Generic Binding, Kodein uses a function `typeToken` that returns either a `Class` or a `ParameterizedType`.
It llooks like on Android 4.4 and before, ParameterizedTypeImpl does not implements equals and hashcode and therefore: `typeToken<Bar<Foo>> != typeToken<Bar<Foo>>`, which causes the bug.

I'm working on a solution, It should land tomorow ;)

**Comment by SalomonBrys** [comment]
This is corrected in version 1.3.0.

Thanks a lot for this report ;)

**Comment by k-kagurazaka** [comment]
Great work:+1:
